server/routes/serviceProviderReferrals/serviceProviderReferralsController.ts

```diff
@@ -629,4 +633,76 @@ export default class ServiceProviderReferralsController {
 
     res.redirect(303, `/service-provider/end-of-service-report/${draftEndOfServiceReport.id}/outcomes/1`)
   }
+
+  async editEndOfServiceReportOutcome(req: Request, res: Response): Promise<void> {
+    const { accessToken } = res.locals.user.token
+
+    const endOfServiceReport = await this.interventionsService.getEndOfServiceReport(accessToken, req.params.id)
+    const referral = await this.interventionsService.getSentReferral(accessToken, endOfServiceReport.referralId)
+
+    const desiredOutcomeNumber = Number(req.params.number)
+
+    if (desiredOutcomeNumber > referral.referral.desiredOutcomesIds.length) {
+      throw createError(404, 'Outcome number is out of bounds')
+    }
+
+    const desiredOutcomeId = referral.referral.desiredOutcomesIds[desiredOutcomeNumber - 1]
+
+    const serviceCategory = await this.interventionsService.getServiceCategory(
+      accessToken,
+      referral.referral.serviceCategoryId
+    )
+
+    const desiredOutcome = serviceCategory.desiredOutcomes.find(val => val.id === desiredOutcomeId)
+
+    if (desiredOutcome === undefined) {
+      throw new Error(`Desired outcome for ID ${desiredOutcomeId} not found`)
+    }
+
+    const outcome = endOfServiceReport.outcomes.find(val => val.desiredOutcome.id === desiredOutcomeId) ?? null
+
+    let formValidationError: FormValidationError | null = null
+    let userInputData: Record<string, unknown> | null = null
+
+    if (req.method === 'POST') {
+      userInputData = req.body
+      const form = new EndOfServiceReportOutcomeForm(req, desiredOutcomeId, referral.referral.serviceUser)
+      const formData = await form.data()
+
+      if (formData.error) {
+        formValidationError = formData.error
+        res.status(400)
+      } else {
+        await this.interventionsService.updateDraftEndOfServiceReport(
+          accessToken,
+          endOfServiceReport.id,
+          formData.paramsForUpdate
+        )
+
+        const isLastDesiredOutcome = desiredOutcomeNumber === referral.referral.desiredOutcomesIds.length
+        if (isLastDesiredOutcome) {
```

Comment:
There's quite a lot of nested `if`s in this statement, which I suppose is a drawback of our combining of the POST / GET functions. I wonder if there's a way we could simplify this in future, see http://wiki.c2.com/?ArrowAntiPattern if interested.

Reply:
Yeah, without actually starting to split it up into smaller methods (which, as it says there, might not have an obvious logical purpose, but also perhaps they might), I'm not sure there's loads to be done here. We already exit early in the POST success case. Perhaps there's an `if` or two that could be turned into a ternary (e.g. where we redirect to).